mtags/src/main/scala/scala/meta/internal/mtags/JavaToplevelMtags.scala

```diff
@@ -0,0 +1,272 @@
+package scala.meta.internal.mtags
+
+import scala.annotation.tailrec
+
+import scala.meta.dialects
+import scala.meta.inputs.Input
+import scala.meta.inputs.Position
+import scala.meta.internal.semanticdb.Language
+import scala.meta.internal.semanticdb.SymbolInformation
+import scala.meta.internal.tokenizers.CharArrayReader
+import scala.meta.internal.tokenizers.Chars._
+import scala.meta.internal.tokenizers.Reporter
+
+class JavaToplevelMtags(val input: Input.VirtualFile) extends MtagsIndexer {
+
+  import JavaToplevelMtags._
+
+  val reporter: Reporter = Reporter(input)
+  val reader: CharArrayReader =
+    new CharArrayReader(input, dialects.Scala213, reporter)
+
+  override def language: Language = Language.JAVA
+
+  override def indexRoot(): Unit = {
+    if (!input.path.endsWith("module-info.java")) {
+      reader.nextRawChar()
+      loop
+    }
+  }
+
+  private def loop: Unit = {
+    val token = fetchToken
+    token match {
+      case Token.EOF =>
+      case Token.Package =>
```

Comment:
Actually, we could stop after reading the package no? You can only have one public class per file and that needs to correspond to the file name. So essentially, when we figure out the package we can stop.

Reply:
Yep, I thought about going this way. However, in this case it won't be possible to search non public symbols and see docs in hover for such classes. It isn't a common case but I remember that there was at least one moment when I wanted to find such class.

Reply:
Will those class on also end up as a candidate for imports, which would be an error?

Reply:
🤔 Probably they won't. We checks if symbol is public one in workspace search for completions

Reply:
When would we need to show scaladocs for private classes though, I think they can't really exist in the code, can they? 

This doesn't compile:
`val a: Testable = new Test().get()`

and a Java file:
```java
package example;

public class Test {
  public Testable get(){
    return new Testable();
  }
}

class Testable {
	public void test() {
		System.out.println("Testable.test()");
	}
}
```

I think adding private classes could actually be confusing.

Reply:
#5024  bring java hovers, so in this case docs search is needed.

Actually, that's the reason why I decided to fix original issue

Reply:
also realized that's it's needed for go-to-definition: non public class might be defined in a different source but in the same package. I hit this case once in dep sources.
```java
/a/src/main/java/example/Test.java
package example;

public class Test {
  public Testable get(){
    return new @@Testable();
  }
}
/a/src/main/java/exmaple/PacakageThings
package example;

public class Smth { ... }

class Testable {
	public void test() {
		System.out.println("Testable.test()");
	}
}
```